### PR TITLE
add pagination and limit for tags pages

### PR DESCRIPTION
**Pull request by syl10100**
Several changes :

- limit results size to 10000 lines for aggregation results like https://fr.openfoodfacts.org/ingredients
- add pagination if results size is over 10000 lines 
- add sub get_cache_results and set_cache_results for memcached usage
- add mongodb count_documents results in memcached

#262
#1630 
#2826

### CODE REVIEW

**Comment by syl10100** [comment]
How it looks like :

OFF_world_ingredients
OFF_world_ingredients_2

**Comment by stephanegigandet** [comment]
Wonderful! Thank you @syl10100 !
I deployed it on the dev server, but for some reason, I don't seem to get the pagination, only the limit for 10000 results.

image

**Comment by stephanegigandet** [comment]
I do get the pagination if I log out.

**Comment by stephanegigandet** [comment]
Overall it seems to be working very well. :)

**Comment by stephanegigandet** [comment]
There's one case for which we want to disable the limit: when there is a ?filter= directive.

e.g. https://fr.openfoodfacts.org/ingredients?filter=oliv

This is used to filter the results based on their value in the local language (so not what's stored in the tags fields). So we have to get all the results, and the filtering is done by the Perl code.

**Comment by syl10100** [comment]
@stephanegigandet  Ok I will check the issue with pagination when you are log in.
For the  ?filter= directive, you want to disable limit for any filter? Because I just tried https://fr.openfoodfacts.org/ingredients?filter=oliv and I get a 504 error (time out)...

**Comment by stephanegigandet** [comment]
> Because I just tried https://fr.openfoodfacts.org/ingredients?filter=oliv and I get a 504 error (time out)...

We had some load issues today.. It's now working again (hopefully). The query takes about 10 second as it fetches all the products.

**Comment by stephanegigandet** [comment]
> For the ?filter= directive, do you want to disable limit for any filter?

@syl10100 : Yes.

File: lib/ProductOpener/Display.pm [hunk 1]
```diff
@@ -1304,7 +1312,42 @@ sub display_mission($)
...
+sub get_cache_results($$){
+
+	my $key = shift;
+	my $request_ref = shift;
+	my $results;
+
+	$log->debug("MongoDB hashed query key", { key => $key }) if $log->is_debug();
+
+	# disable caching if ?nocache=1
+	# or if the user is logged in and nocache is different from 0
```
Comment (syl10100) [inline lib/ProductOpener/Display.pm:1324]: @stephanegigandet is it still relevant to disable cache when user is logged in ?

**Comment by syl10100** [comment]
Correction done for ?filter and pagination for user logged in

**Comment by stephanegigandet** [comment]
Looks good to me. I'm going to make a few small changes. I think we can avoid having a separate count query when there are less than 10k results (which is probably the large majority of the case): we have the results, Perl can easily count them.

**Pull request merged**